resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/DriverInitContainerComponentsProvider.scala

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.kubernetes.submit.v2
+
+import org.apache.spark.{SecurityManager, SparkConf}
+import org.apache.spark.deploy.kubernetes.{SparkPodInitContainerBootstrap, SparkPodInitContainerBootstrapImpl, SubmittedDependencyInitContainerVolumesPluginImpl}
+import org.apache.spark.deploy.kubernetes.config._
+import org.apache.spark.deploy.kubernetes.constants._
+import org.apache.spark.deploy.rest.kubernetes.v2.RetrofitClientFactoryImpl
+
+/**
+ * Interface that wraps the provision of everything the submission client needs to set up the
+ * driver's init-container. This is all wrapped in the same place to ensure that related
+ * components are being constructed with consistent configurations with respect to one another.
+ */
+private[spark] trait DriverInitContainerComponentsProvider {
+
+  def provideInitContainerConfigMapBuilder(
+      maybeSubmittedResourceIds: Option[SubmittedResourceIds])
+      : SparkInitContainerConfigMapBuilder
+  def provideContainerLocalizedFilesResolver(): ContainerLocalizedFilesResolver
+  def provideExecutorInitContainerConfiguration(): ExecutorInitContainerConfiguration
+  def provideInitContainerSubmittedDependencyUploader(
+      driverPodLabels: Map[String, String]): Option[SubmittedDependencyUploader]
+  def provideSubmittedDependenciesSecretBuilder(
+      maybeSubmittedResourceSecrets: Option[SubmittedResourceSecrets])
+      : Option[SubmittedDependencySecretBuilder]
+  def provideInitContainerBootstrap(): SparkPodInitContainerBootstrap
+}
+
+private[spark] class DriverInitContainerComponentsProviderImpl(
+    sparkConf: SparkConf,
+    kubernetesAppId: String,
+    sparkJars: Seq[String],
+    sparkFiles: Seq[String])
```

Comment:
`Client` only uses it for validation that duplicate file names are not present, however. It's unclear if this validation should be done elsewhere.